### PR TITLE
[Metricbeat] Create ActiveMQ module

### PR DESCRIPTION
**Pull request by mtojek**
_Marking as WIP to let you review changes before I dive into testing and dashboards._

Issue: https://github.com/elastic/beats/issues/14510

This PR introduces new Metricbeat module supporting ActiveMQ. The module collects metrics based on Jolokia REST endpoint.

As this is a lightweight module, I believe I can omit integration tests.

still TODO in this PR:
- system test to verify if metrics are collected based on interaction with ActiveMQ
- system test to verify supported versions

### CODE REVIEW

File: x-pack/metricbeat/docker-compose.yml [hunk 1]
```diff
@@ -12,6 +12,16 @@ services:
...
   # Modules
+  activemq:
+    image: docker.elastic.co/observability-ci/beats-integration-activemq:${ACTIVEMQ_VERSION:-5.15.9}-1
+    build:
+      context: ./module/activemq/_meta
+      args:
+        ACTIVEMQ_VERSION: ${ACTIVEMQ_VERSION:-5.15.9}
+    ports:
+      - 8161
+      - 61613
```
Comment (exekias) [inline x-pack/metricbeat/docker-compose.yml]: is this port needed somewhere?

File: x-pack/metricbeat/metricbeat.reference.yml [hunk 2]
```diff
@@ -131,6 +131,15 @@ metricbeat.modules:
...
   #diskio.include_devices: []
 
+#------------------------------- Activemq Module -------------------------------
+- module: activemq
+  metricsets: ['broker', 'queue', 'topic']
+  period: 10s
+  hosts: ['localhost:8161']
+  path: '/api/jolokia/?ignoreErrors=true&canonicalNaming=false'
+  username: admin # default username
+  password: admin # default password
```
Comment (exekias) [inline x-pack/metricbeat/metricbeat.reference.yml]: I understand these are the actual defaults? this should be commented out if not

File: x-pack/metricbeat/module/activemq/broker/_meta/docs.asciidoc [hunk 3]
```diff
@@ -0,0 +1,12 @@
+Broker Metricset retrieves the JMX key `org.apache.activemq:brokerName=*,type=Broker` using Jolokia. It exposes the following metrics:
+
+* `name`: Broker name.
```
Comment (exekias) [inline x-pack/metricbeat/module/activemq/broker/_meta/docs.asciidoc]: you don't need to write this list here, as it will be linked automatically from this page. See https://www.elastic.co/guide/en/beats/metricbeat/current/metricbeat-metricset-kafka-consumergroup.html as an example. It links to https://www.elastic.co/guide/en/beats/metricbeat/current/exported-fields-kafka.html

File: x-pack/metricbeat/module/activemq/broker/_meta/docs.asciidoc [hunk 4]
```diff
@@ -0,0 +1,12 @@
+Broker Metricset retrieves the JMX key `org.apache.activemq:brokerName=*,type=Broker` using Jolokia. It exposes the following metrics:
```
Comment (exekias) [inline x-pack/metricbeat/module/activemq/broker/_meta/docs.asciidoc]: Could you extend this description a little bit? More than the underlying retrieval method, users will be also interested in know what is it about. ie Explain this will retrieve metrics from the brokers

File: x-pack/metricbeat/tests/system/test_activemq.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+    def get_activemq_module_config(self, metricset):
+        return {
+            "name": "activemq",
+            "metricsets": [metricset],
+            "period": "5s",
+            "hosts": self.get_hosts(),
+            "path": "/api/jolokia/?ignoreErrors=true&canonicalNaming=false",
+            "username": "admin",
+            "password": "admin"
```
Comment (exekias) [inline x-pack/metricbeat/tests/system/test_activemq.py]: ok I see it's actually needed

File: x-pack/metricbeat/docker-compose.yml (hunk 1 above)
Comment (mtojek) [reply x-pack/metricbeat/docker-compose.yml]: 8161 - REST endpoint, also Jolokia API exposed
61613 - STOMP, used by system tests to create queue and topic

File: x-pack/metricbeat/metricbeat.reference.yml (hunk 2 above)
Comment (mtojek) [reply x-pack/metricbeat/metricbeat.reference.yml]: The default installation of ActiveMQ uses admin/admin credentials. I wouldn't like to modify the out-of-the-box setup.

File: x-pack/metricbeat/tests/system/test_activemq.py (hunk 5 above)
Comment (mtojek) [reply x-pack/metricbeat/tests/system/test_activemq.py]: yup

File: x-pack/metricbeat/module/activemq/broker/_meta/docs.asciidoc (hunk 3 above)
Comment (mtojek) [reply x-pack/metricbeat/module/activemq/broker/_meta/docs.asciidoc]: Fixed.

File: x-pack/metricbeat/module/activemq/broker/_meta/docs.asciidoc (hunk 4 above)
Comment (mtojek) [reply x-pack/metricbeat/module/activemq/broker/_meta/docs.asciidoc]: I rephrased this paragraph. Feel free to review it!

File: x-pack/metricbeat/tests/system/test_activemq.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+            'name': 'activemq',
+            'metricsets': [metricset],
+            'period': '5s',
+            'hosts': self.get_hosts(),
+            'path': '/api/jolokia/?ignoreErrors=true&canonicalNaming=false',
+            'username': 'admin',
+            'password': 'admin'
+        }
+
+    def get_stomp_host_port(self):
```
Comment (exekias) [inline x-pack/metricbeat/tests/system/test_activemq.py]: I wonder why did you need this instead of `compose_host`? I guess this is similar to https://github.com/elastic/beats/pull/14609

If that's the case, it probably makes sense to push it to compose.py

 cc @jsoriano

Comment (mtojek) [reply x-pack/metricbeat/tests/system/test_activemq.py]: In this case I need to expose two kind of ports. First - REST endpoint with Jolokia and second one - STOMP to communicate with server. Method `compose_host` returns first exposed port of the container. Unfortunately STOMP one is the second.

Comment (jsoriano) [reply x-pack/metricbeat/tests/system/test_activemq.py]: @mtojek `compose_host` can be parameterized with the service and the port you want to get for scenarios with multiple services or multiple ports. In this case for multiple ports you can use `self.compose_host(port='tcp/61613')`. This is used in `test_haproxy.py` for example.

Comment (mtojek) [reply x-pack/metricbeat/tests/system/test_activemq.py]: Thanks for the suggestion. Fixed!

**Comment by exekias** [comment]
ey Marcin, is this one still in WIP? I think you can drop that from the title already :smile_cat: Also, we have the `in_progress` label to flag that if you prefer

**Comment by mtojek** [comment]
You're right @exekias ! Totally forgot about it.

**Comment by mtojek** [comment]
Just rebased against latest master

File: metricbeat/docs/modules/activemq.asciidoc [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+This module periodically fetches JMX metrics from Apache ActiveMQ.
+
+[float]
+=== Compatibility
+The module has been tested with ActiveMQ 5.13.0 and 5.15.9. Other versions are expected to work.
+
+[float]
+=== Usage
+The ActiveMQ module requires <<metricbeat-module-jolokia,Jolokia>>to fetch JMX metrics. Refer to the link for instructions about how to use Jolokia.
```
Comment (kaiyan-sheng) [inline metricbeat/docs/modules/activemq.asciidoc]:
```suggestion
The ActiveMQ module requires <<metricbeat-module-jolokia,Jolokia>> to fetch JMX metrics. Refer to the link for instructions about how to use Jolokia.
```

**Comment by mtojek** [comment]
@kaiyan-sheng

Thanks! Changelog.next has been updated.

**Comment by mtojek** [comment]
As the PR got approved and Travis build passed, I will merge it to master.
